sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotThrowFromDestructors.cs

```diff
@@ -0,0 +1,45 @@
+﻿using System;
+
+namespace Tests.Diagnostics
+{
+    class Program : IDisposable
+    {
+        public void Foo()
+        {
+            throw new Exception(); // Compliant
+        }
+
+        public Program()
+        {
+            throw new Exception(); // Compliant, ctor covered by S3877
+        }
+
+        ~Program()
+        {
+            throw new Exception(); // Noncompliant {{Remove this 'throw' statement.}}
+//          ^^^^^^^^^^^^^^^^^^^^^^
+
+            void Inner()
+            {
+                throw new Exception(); // Noncompliant, C# 7+
+            };
+
+            try
+            {
+                throw new Exception(); // Noncompliant, generally a bad idea to throw and catch in the same method
+            }
+            catch (Exception)
+            {
+            }
+
+            try
+            {
+                Foo();
+            }
+            catch (Exception)
+            {
+                throw; // Compliant, it was already thrown, we are not supposed to swallow all exceptions
```

Comment:
Not 100% sure about this, would love to hear comments.

Reply:
Interesting. I agree with your comment in the test - _rethrow_ is not the same as _throw_. It's more like the destructor has decided not to squash those exceptions it was expecting.

Reply:
I am fine with both approaches (raising and not raising).